staging/src/k8s.io/client-go/informers/factory.go

```diff
@@ -137,6 +149,52 @@ func (f *sharedInformerFactory) Start(stopCh <-chan struct{}) {
 	}
 }
 
+// informerStopped removes a stopped informer from the factory's list
+// of informers and started informers.
+func (f *sharedInformerFactory) informerStopped(informerType reflect.Type) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	delete(f.startedInformers, informerType)
+	delete(f.informers, informerType)
+}
+
+// StartWithStopOptions initializes all requested informers with the stop options provided, defaulting to
+// the old stop options (only stopping via closure of stopCh).
+// It makes sure remove an informer from the list of informers and started informers when the informer is stopped
+// to prevent a race where InformerFor gives the user back a stopped informer that will never be started again.
+func (f *sharedInformerFactory) StartWithStopOptions(stopCh <-chan struct{}) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	for informerType, informer := range f.informers {
+		if !f.startedInformers[informerType] {
+			var stopOptions cache.StopOptions
+			if opts, ok := f.customStopOptions[informerType]; ok {
+				// TODO: is it safe to overwrite the externalStop?
+				// How do we ensure creator of stopopts isn't adding a stopCh that they expect to be used?
+				// Should we merge them instead?
+				stopOptions = opts
+
+			} else {
+				// default to old behavior of never stopping.
+				stopOptions = cache.StopOptions{
+					OnListError: func(err error) bool {
+						return false
+					},
+				}
+			}
+			stopOptions.ExternalStop = stopCh
```

Comment:
this is gonna clobber the passed-in ExternalStop, right?  If so, the `WithCustomStopOptions` shouldn't take a `StopOptions` struct, since one of the fields is not intended to be set.

Reply:
Got rid of the  `WithCustomStopOptions` in favor of a single policy for the entire informer (configured via `WithOnListError`).